servant-foreign/src/Servant/Foreign/Internal.hs

```diff
@@ -120,6 +122,9 @@ type family Elem (a :: *) (ls::[*]) :: Constraint where
   Elem a (a ': list) = ()
   Elem a (b ': list) = Elem a list
 
+class HasForeignType a where
+    typeFor :: Proxy a -> ForeignType
```

Comment:
One issue with this is that there'll have to be orphan instances. Also, the class `HasForeignTypes` will mean different things in different contexts, and things will only work if instances for different languages aren't ever in scope at the same time. Would it not perhaps be better to have another type parameter?

```
class HasForeignType typ lang where
     typeFor :: Proxy lang -> Proxy a -> ForeignType
```

That way each package can declare its own `lang` datatype (e.g. `data JSLang`) in its own package, and different instances won't conflict across languages.